### PR TITLE
WIP: make systemd code base cert-API01-C compliant

### PR DESCRIPTION
**Pull request by FrazerClews**
https://wiki.sei.cmu.edu/confluence/display/c/API01-C.+Avoid+laying+out+strings+in+memory+directly+before+sensitive+data

### CODE REVIEW

**Comment by evverx** [comment]
Given that instead of proactively looking for buffer overruns with fuzz targets or tests run under ASan it kind of basically recommends sweeping them under the rug and hoping for the best I'd say it's the most dubious recommendation I've seen there.

@FrazerClews since it seems to be the second PR you've opened where the codebase is supposed to be made more secure, I'd like to point out that I think it would be great (and probably much more helpful) if you could take a look at the coverity report at https://scan.coverity.com/projects/350 or at the LGTM report at https://lgtm.com/projects/g/systemd/systemd/alerts/?mode=list and help to address issues reported there. Other than that, we have a bunch of fuzzers covering https://oss-fuzz.com/coverage-report/job/libfuzzer_asan_systemd/latest. If you could improve the fuzz targets or add new ones I'm pretty sure they would be much more effective and efficient in terms of preventing issues that you're trying to prevent as far as I can tell.

**Comment by evverx** [comment]
Other than that, I'm not sure what for example https://github.com/systemd/systemd/pull/16123/files#diff-67a7a6fb13bf7e5adfe383aed8a05ebe (where an array containing two file descriptors was moved) is trying to accomplish. As far as I can see it has nothing to do with cert-API01-C. Was it changed automatically?

File: src/boot/efi/linux.h [hunk 1]
```diff
@@ -46,33 +46,33 @@ struct setup_header {
 
 /* adapted from linux' bootparam.h */
 struct boot_params {
+        UINT64 tboot_addr;
```
Comment (poettering) [inline src/boot/efi/linux.h:49]: we cannot reorder this, the order has to match what linux passes in to us

File: src/boot/efi/pe.c [hunk 2]
```diff
@@ -7,7 +7,6 @@
 #include "util.h"
 
 struct DosFileHeader {
-        UINT8   Magic[2];
```
Comment (poettering) [inline src/boot/efi/pe.c:10]: we cannot reorder this, the order has to match what is in a PE executable header

File: src/boot/efi/splash.c [hunk 3]
```diff
@@ -8,10 +8,10 @@
 #include "util.h"
 
 struct bmp_file {
-        CHAR8 signature[2];
         UINT32 size;
-        UINT16 reserved[2];
         UINT32 offset;
+        CHAR8 signature[2];
+        UINT16 reserved[2];
```
Comment (poettering) [inline src/boot/efi/splash.c:14]: as above. the BMP file header has an order

**Pull request closed without merging**